### PR TITLE
Use CAPV util function to retrieve VSphere cluster credentials.

### PR DESCRIPTION
**Pull request by HanFa**
### What this PR does / why we need it

To retrieve the credential from a VSphereCluster, instead of just getting it from the associated secret object, we use CAPV util function to cover the case of multi-tenancy. 

### Which issue(s) this PR fixes

Fixes #

https://github.com/vmware-tanzu/tanzu-framework/issues/3392

### Describe testing done for PR

Addon Controller integration test with `KUBEBUILDER_ASSETS=/Users/hanfa/workspace/tanzu-framework-HanFa/hack/tools/bin/kubebuilder/bin go test ./controllers -v`

Manually testing for non-paravirt uTKG

### Release note

```release-note
package-based-lcm: Use CAPV util function to retrieve VSphere cluster credentials.
```

### CODE REVIEW

**Review by lubronzhan — APPROVED** [review]
lgtm

**Comment by randomvariable** [comment]
Split the testutil changes into a separate commit, otherwise lgtm

File: addons/controllers/vspherecpiconfig_controller_test.go [hunk 1]
```diff
@@ -268,6 +271,49 @@ var _ = Describe("VSphereCPIConfig Reconciler", func() {
...
+			Eventually(func() bool {
+				secretKey := client.ObjectKey{
+					Namespace: clusterNamespace,
+					Name:      fmt.Sprintf("%s-%s-data-values", clusterName, constants.CPIAddonName),
+				}
+				if err := k8sClient.Get(ctx, secretKey, secret); err != nil {
+					return false
+				}
+				secretData := string(secret.Data["values.yaml"])
+				fmt.Println(secretData) // debug dump
```
Comment (lubronzhan) [inline addons/controllers/vspherecpiconfig_controller_test.go:307]: Remove this

File: addons/controllers/vspherecpiconfig_controller_test.go [hunk 2]
```diff
@@ -268,6 +271,49 @@ var _ = Describe("VSphereCPIConfig Reconciler", func() {
...
+				Expect(len(secretData)).Should(Not(BeZero()))
```
Comment (lubronzhan) [inline addons/controllers/vspherecpiconfig_controller_test.go:308]: nit: shouldnot

**Review by lubronzhan — APPROVED** [review]

File: addons/controllers/vspherecpiconfig_controller_test.go (hunk 2 above)
Comment (HanFa) [reply addons/controllers/vspherecpiconfig_controller_test.go]: fixed

File: addons/controllers/vspherecpiconfig_controller_test.go [hunk 3]
```diff
@@ -250,6 +253,48 @@ var _ = Describe("VSphereCPIConfig Reconciler", func() {
...
+					Namespace: clusterNamespace,
+					Name:      fmt.Sprintf("%s-%s-data-values", clusterName, constants.CPIAddonName),
+				}
+				if err := k8sClient.Get(ctx, secretKey, secret); err != nil {
+					return false
+				}
+				secretData := string(secret.Data["values.yaml"])
+				Expect(len(secretData)).ShouldNot(BeZero())
+				Expect(strings.Contains(secretData, "username: administrator@vsphere.local")).Should(BeTrue())
+				Expect(strings.Contains(secretData, "password: Admin!23")).Should(BeTrue())
```
Comment (maralavi) [inline addons/controllers/vspherecpiconfig_controller_test.go:291]: Maybe better not to commit these credentials here?

Comment (HanFa) [reply addons/controllers/vspherecpiconfig_controller_test.go]: thanks, just fixed

**Comment by maralavi** [comment]
@HanFa Will a resource of kind VSphereClusterIdentity or an identityRef on a VsphereCluster exist for sure in a cluster?  How credentials would be fetched if they don’t?

**Comment by HanFa** [comment]
> Will a resource of kind VSphereClusterIdentity or an identityRef on a VsphereCluster exist for sure in a cluster? How credentials would be fetched if they don’t?

Hi @maralavi , not necessarily. If the  VSphereClusterIdentity doesn't exist, the secretRef type should be `Secret` instead, the controller shall grab the credentials from the ref-ed secret object.

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 4]
```diff
@@ -95,15 +97,13 @@ func (r *VSphereCPIConfigReconciler) mapCPIConfigToDataValuesNonParavirtual( //
 	d.TLSThumbprint = vsphereCluster.Spec.Thumbprint
 	d.Server = vsphereCluster.Spec.Server
 
-	// derive vSphere username and password from the <cluster name> secret
-	clusterSecret, err := r.getSecret(ctx, cluster.Namespace, cluster.Name)
+	// derive vSphere username and password using CAPV util function
+	clusterCredentials, err := capvidentity.GetCredentials(ctx, r.Client, vsphereCluster, capvmanager.DefaultPodNamespace)
```
Comment (maralavi) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:101]: Can we expect that this ```DefaultPodNamespace``` always get populated correctly?

Comment (HanFa) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:101]: no, the DefaultPodNamespace is always `capv-system`, we can't gurantee its existence. if it doesn't exist, the `GetCredentials` will just use the cluster's namespace

**Comment by HanFa** [comment]
> > > Will a resource of kind VSphereClusterIdentity or an identityRef on a VsphereCluster exist for sure in a cluster? How credentials would be fetched if they don’t?
> > 
> > 
> > Hi @maralavi , not necessarily. If the VSphereClusterIdentity doesn't exist, the secretRef type should be `Secret` instead, the controller shall grab the credentials from the ref-ed secret object.
> 
> Thanks for the clarification @HanFa . Can we expect that `capvidentity.GetCredentials()` take care of that case too? (I ask as now `getSecret()` is removed?)

yes, that's exactly the case

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/9492c4bb01f03a0cec534311ed1d61515e94eddf/pkg/identity/identity.go#L53

**Review by maralavi — APPROVED** [review]

**Pull request merged**